### ISSUE TITLE
[HOLD for payment 2023-10-25] [$500] Web - The horizontal line separating the header and the messages in the IOU report room disappears and reappears irregularly

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#26529]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
- You Can open any IOU report room and skip the first two steps
1. Click on the FAB and request Money from any user this will take you to that chat room
2. Click on the IOU message to open the IOU report room
3. Hover over the first Message in the IOU report room and mark it as unread
4. Now notice the unread mark appears and the grey horizontal line still appears
5. leave the the IOU report room and return
6. Notice the horizontal grey line disappears
7. Hover over the first message again and mark it as unread again
8. Now the horizontal grey line appears again

## Expected Result:
The unread indicator should be consistently be aligned with the grey bar in IOUs/tasks.

## Actual Result:
The horizontal line separating the header and the messages in the IOU report room disappears and reappears irregularly.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.61.0
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

[Screencast from 28-08-23 02_23_41.webm](https://github.com/Expensify/App/assets/93399543/74b0d598-f818-4a36-92a5-317eed84c90a)

https://github.com/Expensify/App/assets/93399543/5a24d438-2352-4f63-b4bc-a4010d95477e

**Expensify/Expensify Issue URL:**
**Issue reported by:** @Ahmed-Abdella
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1693183486459189

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01a365a66bcfc380e7</li>
        <li>Upwork Job ID: 1699177381705969664</li>
        <li>Last Price Increase: 2023-09-05</li>
<li>Automatic offers: </li>
<ul>
<li>Ahmed-Abdella | Reporter | 26638143</li>
</ul></li>
    </ul>
</details>

**Comment by joekaufmanexpensify** [comment]
OOO today, will triage tomorrow!

**Comment by joekaufmanexpensify** [comment]
IMO, the behavior of the unread indicator here is consistent with the rest of the app. We just need to figure out the alignment of the unread bar.

**Comment by joekaufmanexpensify** [comment]
Or alternatively, perhaps there shouldn't be margin between the top the first message in the IOU/task room and the grey bar.

**Comment by zukilover** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
The unread indicator should be consistently be aligned with the grey bar in IOUs/tasks.

### What is the root cause of that problem?
This is caused by the `shouldHideThreadDividerLine` flag not being read correctly by `ReportActionItem` component. We use `useCallback` to render the item, and the `shouldHideThreadDividerLine` is evaluated within the callback and not as a dependency. This caused the `ReportActionItem` to not re-render when it should.

https://github.com/Expensify/App/blob/1bdc5acc12f982b4f6e37bfadf9060ac3d2d3b3c/src/pages/home/report/ReportActionsList.js#L249-L250

Changing the style so that the unread indicator overlaps the grey line is not the best solution because it's not addressing the underlying problem. Also, the unread indicator has `new` label on the right that uses transparent background. A portion of the grey line will be exposed if we only use the styling fix.

I propose the following:

Change `currentUnreadMarker` from `useRef` to `useState` so that it triggers re-render when it should.
Change `shouldHideThreadDividerLine` to `useMemo` and add it as dependency to the renderItem.

### What changes do you think we should make in order to solve the problem?
```
    const [currentUnreadMarker, setCurrentUnreadMarker] = useState(null);
    const shouldHideThreadDividerLine = useMemo(() => {
        return sortedReportActions.length > 1 && sortedReportActions[sortedReportActions.length - 2].reportActionID === currentUnreadMarker
    }, [sortedReportActions, currentUnreadMarker])
```

### What alternative solutions did you explore? (Optional)
N/A

**Comment by joekaufmanexpensify** [comment]
@robertKozik mind taking a look at this proposal when you have a sec?

**Comment by robertKozik** [comment]
Yeah I'm on it 👀

**Comment by robertKozik** [comment]
Sorry for keeping you waiting @zukilover, I think your proposal is good, I checked it briefly on my branch and it solved the problem. Let's see it in action!

Selected proposal: [Proposal](https://github.com/Expensify/App/issues/26529#issuecomment-1707444942)
Contributor: @zukilover

🎀 👀 🎀 C+ reviewed

**Comment by joekaufmanexpensify** [comment]
Pending internal engineering review.

**Comment by joekaufmanexpensify** [comment]
@deetergp mind signing off on this proposal when you have a sec?

**Comment by deetergp** [comment]
I agree with @robertKozik, let's go with @zukilover's proposal 👍

**Comment by zukilover** [comment]
I have applied for the job on Upwork. PR will be ready in approx. 3 hrs.

**Comment by zukilover** [comment]
Hi @joekaufmanexpensify , I'm new to this, but I don't see any offer on Upwork. How should I proceed?

**Comment by Ahmed-Abdella** [comment]
@zukilover I think he will send you the offer after your application

**Comment by zukilover** [comment]
I haven't got the offer on Upwork yet, but the PR is already there. It'd be appreciated if you could take a look @joekaufmanexpensify

**Comment by joekaufmanexpensify** [comment]
@zukilover did you apply to the upwork job (per [this message](https://github.com/Expensify/App/issues/26529#issuecomment-1716272373))?

**Comment by zukilover** [comment]
Yes, I have applied to the job @joekaufmanexpensify

**Comment by zukilover** [comment]
Hi @joekaufmanexpensify is there any updates on the issue above?

It's happening to my other proposal as well: https://github.com/Expensify/App/issues/27021
I haven't got the offer for that either.

**Comment by joekaufmanexpensify** [comment]
@zukilover Thanks for following up. Sometimes we might not send the upwork offer right when you are assigned the job on GH. But we always will prior to when the issue is due for payment. I've just accepted your offer on Upwork for this job.

**Comment by robertKozik** [comment]
Checking it 👀

**Comment by robertKozik** [comment]
I believe you're right @dukenv0307 it still can be reproduced when creating the task. During my investigation I found out that it only occurred for the first access to the task (only after user got redirected after creating one), in other cases horizontal line behaves nicely. Can you check it as well @zukilover?

**Comment by Ahmed-Abdella** [comment]
@robertKozik May I suggest something? I believe that the task room can be fixed similarly to the IOU room without creating a separate issue. Additionally, I have noticed that the unread mark does not function perfectly after being fixed. It is moving from bottom to top, which it should not be, and a black horizontal line briefly appears after removing the gray horizontal line. Therefore, I suggest that we properly fix the issue and apply the same solution to the task report.

[Screencast from 20-09-23 15:22:34.webm](https://github.com/Expensify/App/assets/77763699/690d6529-7149-4bf4-9eda-8a24643a4007)

**Comment by Ahmed-Abdella** [comment]
It should behave like the normal threads room

[Screencast from 20-09-23 15:39:24.webm](https://github.com/Expensify/App/assets/77763699/07176e7e-fa57-436c-90bc-72ef6ab89f23)

**Comment by zukilover** [comment]
This happens intermittently to me.  
[intermittent.webm](https://github.com/Expensify/App/assets/371291/5eae4b46-cd2b-4702-84fd-0b400fe1db80)


Also @Ahmed-Abdella I don't think this should behave like the normal threads room you've shown above, this is different and was done to handle an unread marker for chat that is exactly below the header separator. What you showed does not have any header separator.

**Comment by thienlnam** [comment]
Repeating what QA mentioned, but the issue still appears to be happening in iOS

**Comment by zukilover** [comment]
@thienlnam can you send a screen record? It works for me:

https://github.com/Expensify/App/assets/371291/fde06e67-99ba-4741-a515-0672c7b9b489

**Comment by thienlnam** [comment]
@kavimuru Do you have more failure details about regarding this issue?

**Comment by deetergp** [comment]
@robertKozik I think we should go ahead and fix tasks here since we've got a solution that will likely work from IOUs.

**Comment by robertKozik** [comment]
@zukilover per @deetergp comment, can you extend implementation of your fix into the task page?

**Comment by zukilover** [comment]
Sure, I'll work on it. @robertKozik

**Comment by zukilover** [comment]
Please review the updates: https://github.com/Expensify/App/pull/28146 @robertKozik

**Comment by zukilover** [comment]
Ready for your approval @deetergp : https://github.com/Expensify/App/pull/28146

**Comment by joekaufmanexpensify** [comment]
Catching up on this one now, and is this a correct understanding of where we're at? We had the first PR for this issue, which fixed the bug in some, but not all situations. Now the PR above, is in review, and should fix it in the remaining situations. 

Is that right?

**Comment by zukilover** [comment]
@joekaufmanexpensify that's about right. The original problem has been solved and completed. The new PR is an extended implementation.

**Comment by joekaufmanexpensify** [comment]
PR still in review

**Comment by joekaufmanexpensify** [comment]
@robertKozik mind completing the BZ checklist here so we can prep to close this one?

**Comment by robertKozik** [comment]
> * [X]  [@robertKozik] The PR that introduced the bug has been identified. Link to the PR: **It had more feature-like vibe, as it was consistent behaviour which needed to be changed ([context](https://github.com/Expensify/App/issues/26529#issuecomment-1707342497))**
> * [X]  [@robertKozik] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: **N/A**
> * [X]  [@robertKozik] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: **N/A**
> * [X]  [@robertKozik] Determine if we should create a regression test for this bug. **I think we don't need it for this one**
> * [X]  [@robertKozik] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again. **N/A**

**Comment by joekaufmanexpensify** [comment]
Great, ty!

**Comment by joekaufmanexpensify** [comment]
BZ checklist is complete, so we need to now issue payment. Here's a summary of payments: 

- @zukilover - $500 for PR. Paid via Upwork. 
- @Ahmed-Abdella $50 for reporting. Paid via upwork. 
- @robertKozik - payment for C+ handled separately.

**Comment by joekaufmanexpensify** [comment]
@zukilover your $500 was already previously paid out from when you submitted your work for approval in Upwork. I ended your contract!

**Comment by joekaufmanexpensify** [comment]
@Ahmed-Abdella $50 sent and contract ended!

**Comment by joekaufmanexpensify** [comment]
Also noting that the initial PR was completed in less than 9 days, which is why no urgency penalty applied here. We later than decided to do a second clean up PR, which was also merged in less than 9 days.

**Comment by joekaufmanexpensify** [comment]
Bug is fixed, BZ checklist complete, and all payment issued. Closing as this is all set. Thanks everyone!
